std/dynamic_library.zig

```diff
@@ -260,21 +279,70 @@ pub const WindowsDynLib = struct {
         self.* = undefined;
     }
 
-    pub fn lookup(self: *WindowsDynLib, name: []const u8) ?usize {
-        return @ptrToInt(windows.kernel32.GetProcAddress(self.dll, name.ptr));
+    pub fn lookupC(self: *WindowsDynLib, comptime T: type, name: [*]const u8) ?T {
+        if (windows.kernel32.GetProcAddress(self.dll, name)) |addr| {
+            return @ptrCast(T, addr);
+        } else {
+            return null;
+        }
+    }
+
+    pub fn lookup(self: *DlDynlib, comptime T: type, comptime max_name_len: usize, name: []const u8) ?T {
+        const c_name: [max_name_len]u8 = undefined;
+        mem.copy(&c_name, name);
+        c_name[name.len] = 0;
+        return self.lookupC(T, &c_name);
+    }
+};
+
+pub const DlDynlib = struct {
+    pub const Error = error{FileNotFound};
+
+    handle: *c_void,
+
+    pub fn open(path: []const u8) !DlDynlib {
+        if (!builtin.link_libc and !os.darwin.is_the_target) {
+            @compileError("DlDynlib requires libc");
+        }
+
+        return DlDynlib{
+            .handle = system.dlopen(path.ptr, system.RTLD_LAZY) orelse {
+                return error.FileNotFound;
+            },
+        };
+    }
+
+    pub fn close(self: *DlDynlib) void {
+        _ = system.dlclose(self.handle);
+        self.* = undefined;
+    }
+
+    pub fn lookupC(self: *DlDynlib, comptime T: type, name: [*]const u8) ?T {
+        if (system.dlsym(self.handle, name)) |symbol| {
```

Comment:
I just recalled something from https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66826
> Now dlsym (and other dl-functions) secretly take shadow parameter - return address on stack

I wonder if we need to reflect this somehow?

Reply:
from what i understand, this seems to only affect tail calls like line 309, though i'm not sure of that would even trigger the issue in that bug report.

Reply:
I don't think zig makes any promises in terms of not tail calling something. The function probably needs some sort of annotation?

Reply:
llvm has a [`notail`](https://llvm.org/docs/LangRef.html#call-instruction) annotation for function calls. this could be a possible use case for a `@call()` builtin. we could have an enum for `Inline`, `NoInline`, `MustTail`, and `NoTail`, and unify the builtin with `@newStackCall`.

Reply:
I think we should have the noinline call also guarantee no tail call. In all use cases where one wants no inline, no tail is also desired.

Reply:
I also would like to allow noinline (with the implication of notail) as part of a function prototype. This way functions such as dlsym can properly  document how to call them, instead of being footguns at the callsites.

Reply:
Created https://github.com/ziglang/zig/issues/3732 to track

Reply:
This is finally unblocked with the merge of #3856